### DISCUSSION TITLE
limit/offset not work while topK is set & limit/offset only work while no search_params is set (Milvus 2.3.1, milvus2-sdk-node 2.3.2)

### DISCUSSION THREAD

**Discussion by grimmerk** [milvus-io/milvus#27696] [Q&A and General discussion] [answered]
Hello, everyone, I’ve encountered some issues and would appreciate any assistance.

### environment:
server: Milvus 2.3.1 in a local docker-compose.
client: @zilliz/milvus2-sdk-node 2.3.2
os: macOs 13.2.1

## Issue1:
while using `serach`,  `limit/offset` parameters do not take effect when `topK` is set in `search_params`. e.g. `limit:3` with `topK:10`. It would return `10` results.

## Issue2:
while using `serach` with `limit/offset` and `search_params` (e.g. `{metric_type: 'L2'}`) set up in the arguments, the response contains `error_code:'UnexpectedError'` and `reason: "topk not found in search_params"`. So if I want to make limit/offset working, I can not set any `search_params`.

**Comment by xiaofan-luan** [comment]
> Hello, everyone, I’ve encountered some issues and would appreciate any assistance.
> 
> ### environment:
> server: Milvus 2.3.1 in a local docker-compose. client: @zilliz/milvus2-sdk-node 2.3.2 os: macOs 13.2.1
> 
> ## Issue1:
> while using `serach`, `limit/offset` parameters do not take effect when `topK` is set in `search_params`. e.g. `limit:3` with `topK:10`. It would return `10`.
> 
> ## Issue2:
> while using `serach` with `limit/offset` and `search_params` (e.g. `{metric_type: 'L2'}`) set up in the arguments, the response contains `error_code:'UnexpectedError'` and `reason: "topk not found in search_params"`. So if I want to make limit/offset working, I can not set any `search_params`.

1. this is expected .it will return top 3 to top 13. if you only 7 result, it should set offset 3, limit7
2. this is a potential bug, could you send your code to reproduce?

**Reply by grimmerk** [reply]
@xiaofan-luan thank you for the quick reply

1. I have tested `topK=10` with the below cases, and they all return the same results array as the 1st case. It seems to mean offset/limit do not take any effect while using topK. 
    1.  no offset/limit
    2. `{limit: 5}`
    3. `{limit: 5, offset: 10}`, 
    4. `{offset: 5}`
2.  Here is the screenshot of my code (either only setting limit or setting limit/offset both happens this issue). Let me know if you need a real working code. If it is needed, we can try to make some example code.

<img width="1221" alt="截圖 2023-10-13 16 28 50" src="https://github.com/milvus-io/milvus/assets/5940941/b222b55a-bbbf-4cca-94f2-654647ee139e">

**Reply by grimmerk** [reply]
BTW, if I omit `search_params` completely and only use offset and limit, what is the default parameter about `anns_field`, `metric_type`, `nprobe` (for `IVF_FLAT`)? I still can get the results, but I'm curious about this detail. My collection has only one vector field, and its index is `IVF_FLAT`. Can it also automatically find out the target vector field?

**Reply by yhmo** [reply]
@grimmer0125 

I guess your parameters are incorrect. The "offset" parameter should be passed within the "params".
This is an example to show how to search with limit/offset parameters:

```
if __name__ == '__main__':
    connections.connect(host='localhost', port='19530')

    field_id = FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False)
    field_vector = FieldSchema(name="vector", dtype=DataType.FLOAT_VECTOR, dim=4)

    collection_name = "aaa"
    if utility.has_collection(collection_name):
        utility.drop_collection(collection_name)

    collection_schema = CollectionSchema(fields=[field_id, field_vector])
    collection = Collection(name=collection_name, schema=collection_schema)

    index_param = {
        "index_type": "FLAT",
        "params": {},
        "metric_type": "L2"
    }
    collection.create_index(field_name='vector', index_params=index_param)

    # insert 5 rows
    data = [
        [1, 2, 3, 4, 5],
        [[1.1, 1.2, 1.3, 1.4], [2.1, 2.2, 2.3, 2.4], [3.1, 3.2, 3.3, 3.4], [4.1, 4.2, 4.3, 4.4], [5.1, 5.2, 5.3, 5.4]],
    ]
    collection.insert(data)
    collection.flush()
    print("Total row count:", collection.num_entities)

    collection.load()

    target_vector = [5.1, 5.2, 5.3, 5.4]
    def search(limit, offset):
        search_params = {"metric_type": "L2", "params": {}, "offset": offset}
        results = collection.search(
            data=[target_vector],
            anns_field="vector",
            param=search_params,
            limit=limit,
            output_fields=['id', 'vector'],
        )
        for result in results:
            for j, res in enumerate(result):
                print("Top {}: {}".format(j, res))

    print("====== search with limit=10, offset=0 ======")
    search(limit=10, offset=0)

    print("====== search with limit=10, offset=3 ======")
    search(limit=10, offset=3)

    print("====== search with limit=3, offset=2 ======")
    search(limit=3, offset=2)
```

The output lines:

```
====== search with limit=10, offset=0 ======
Top 0: id: 5, distance: 0.0, entity: {'vector': [5.1, 5.2, 5.3, 5.4], 'id': 5}
Top 1: id: 4, distance: 4.0, entity: {'vector': [4.1, 4.2, 4.3, 4.4], 'id': 4}
Top 2: id: 3, distance: 16.0, entity: {'vector': [3.1, 3.2, 3.3, 3.4], 'id': 3}
Top 3: id: 2, distance: 36.0, entity: {'vector': [2.1, 2.2, 2.3, 2.4], 'id': 2}
Top 4: id: 1, distance: 64.0, entity: {'vector': [1.1, 1.2, 1.3, 1.4], 'id': 1}
====== search with limit=10, offset=3 ======
Top 0: id: 2, distance: 36.0, entity: {'id': 2, 'vector': [2.1, 2.2, 2.3, 2.4]}
Top 1: id: 1, distance: 64.0, entity: {'id': 1, 'vector': [1.1, 1.2, 1.3, 1.4]}
====== search with limit=3, offset=2 ======
Top 0: id: 3, distance: 16.0, entity: {'id': 3, 'vector': [3.1, 3.2, 3.3, 3.4]}
Top 1: id: 2, distance: 36.0, entity: {'id': 2, 'vector': [2.1, 2.2, 2.3, 2.4]}
Top 2: id: 1, distance: 64.0, entity: {'id': 1, 'vector': [1.1, 1.2, 1.3, 1.4]}
```

**Reply by grimmerk** [reply]
@yhmo thanks for the suggestion. I tried to put offset/limit in `search_params` as suggested, or even put them in `search_params: params:{`, and the results are the same, issue1 and issue2 both exist. Maybe there is some bug on Node.js SDK side?

issue1(limit/offset not work while setting topK): put it in  search_params. Limit/offset does not take effect. 
<img width="496" alt="search_params_level-topK-issue1" src="https://github.com/milvus-io/milvus/assets/5940941/725eb505-aead-4bbe-8d9c-29626ce96f71">

issue1(limit/offset not work while setting topK)): put it in `search_params: params`. Limit/offset does not take effect. 

<img width="519" alt="search_params_params_level-topK-issue1" src="https://github.com/milvus-io/milvus/assets/5940941/15f5220e-e1dc-41a5-9054-1b647931dfa2">

issue2(can not set search_params): put it in  search_params. without topK. Got `topKNotFound`. 
<img width="541" alt="search_params_level-topKNotFound-issue2" src="https://github.com/milvus-io/milvus/assets/5940941/31ba3cf8-4e5d-4062-9916-a57f0c3a332a">

issue2(can not search_params): put it in  `search_params:params`. without topK. Got `topKNotFound`. 
<img width="589" alt="search_params_params_level-topKNotFound-issue2" src="https://github.com/milvus-io/milvus/assets/5940941/67287abb-39ce-4e5f-8671-3e10f53ac35d">

**Reply by nameczz** [reply, marked as answer]
@grimmer0125 search_params is for compatibility with older versions of the SDK. 
source code:
![image](https://github.com/milvus-io/milvus/assets/20559208/d6da9c4d-30cb-42ba-b3e2-5e5c9cf43d20)
So if you both pass search_params, topk and offset.  SDK will use search_params, and ignore topk, offset. 
We do not recommend using search_params, you can code like: 
```
https://milvus.io/api-reference/node/v2.3.x/Data/search.md
new milvusClient(MILUVS_ADDRESS).search({
  collection_name: "my_collection",
  vector: [1, 2, 3, 4],
  filter: "count > 5",
  limit: 10,
  offset: 2,
  metric_type: MetricType.L2,
  param: { nprobe: 1024 },
  consistency_level: ConsistencyLevelEnum.Strong,
  output_fields: ["your field defined in schema"],
});
```

**Reply by grimmerk** [reply]
Thank you @nameczz. It works after making this change.

**Comment by nameczz** [comment]
> BTW, if I omit `search_params` completely and only use offset and limit, what is the default parameter about `anns_field`, `metric_type`, `nprobe` (for `IVF_FLAT`)? I still can get the results, but I'm curious about this detail. My collection has only one vector field, and its index is `IVF_FLAT`. Can it also automatically find out the target vector field?

If anns_field is not passed, sdk will get it from your collection.
